nasa/ow_simulator
# Power fault telemetry unification

## Linked Issues:
| EPIC ⚡ | [OCEANWATER-934](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-934) |
| :----------- | :----------- |
| Jira Ticket 🎟️ | [OCEANWATER-1014](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-1014) |
| Github :octocat:  | [OCEANWATER-1014_implement_telemetry_power_faults_status](https://github.com/nasa/ow_simulator/tree/OCEANWATER-1014_implement_telemetry_power_faults_status) |

Mostly artifact renaming and updating message definitions.

**This issue has a corresponding and required branch on [ow_autonomy](https://github.com/nasa/ow_autonomy/tree/OCEANWATER-1014_implement_telemetry_power_faults_status).**

## Summary of Changes
* Renamed the `PowerFaults` message to `PowerFaultsStatus` and updated all references in source and documentation.
* Moved `PowerFaultsStatus.msg` to the `owl_msgs` package.
* Removed the `/faults/` topic prefix from the `power_faults_status` topic.
* Updated `PowerFaultsStatus` message semantics and defintion.
* Made Plexil-OwInterface package imports consistent with new 'owl_msgs' package.
* Renamed some variables to be consistent with message rename.
* Implemented individual power faults detection and publishing their respective messages.

## Test
* Build according to [User Documentation](https://github.com/nasa/ow_simulator/wiki/Using-OceanWATERS).
* Run four terminals sourced with the project's `setup.bash`.
* In one terminal, run the project:
  * `roslaunch ow atacama_y1a.launch`
* In the second terminal, verify the fault messages are in the `owl_msgs` package and its contents:
  * `rosmsg package owl_msgs`
  * `rosmsg show owl_msgs/PowerFaultsStatus`
* In the second terminal Subscribe to the system faults status topic
  * `rostopic echo /system_faults_status`
* In the third terminal, subscribe to the `/power_faults_status` topic
* In the fourth terminal, subscribe to `/power_system_node/battery_temperature`
* In RQT go to the Dynamic Reconfigure tab and select faults.
* Toggle the high power draw fault (set to 30) and observe the rostopic changes after some time
  - You may wish to increase the timestep in the gazebo tab.
  - You should observe the temperature fault being detected in `/power_faults_status` and a power execution fault being detected in `/system_faults_status`
  - These faults will occur when `/power_system_node/battery_temperature` reports a value > 70.
* Reduce the power draw slider back to the minimum and observe both fault message values return to 0 after some time (when `/power_system_node/battery_temperature` returns to below 70)
* You may continue at high power draw until the battery depletes and you observe the appropriate fault messages.

## Comments/Issues
* The threshold for excessive instantaneous state of charge loss cannot be reached with the current battery model and fault injections, so the respective `/power_faults_status` will not bee seen changing.
